jet of pow using comp with exp, mul, log

jacobjinkelly opened the pull request:
This implementation is numerically unstable (the test begins to fail when the upper limit is about 35), but it's unclear if there's another way to implement jet of pow easily.

Co-authored-by: Jesse Bettencourt <jessebett@cs.toronto.edu> (@jessebett)
Co-authored-by: David Duvenaud <duvenaud@cs.toronto.edu> (@duvenaud)

jacobjinkelly committed:
jet of pow using comp with exp, mul, log

Co-authored-by: Jesse Bettencourt <jessebett@cs.toronto.edu>
Co-authored-by: David Duvenaud <duvenaud@cs.toronto.edu>

shoyer commented on jax/experimental/jet.py:
```diff
@@ -239,6 +239,22 @@ def _expm1_taylor(primals_in, series_in):
 jet_rules[lax.expm1_p] = _expm1_taylor
 
+def _pow_taylor(primals_in, series_in):
+  u_, r_ = primals_in
+  u_terms, r_terms = series_in
+  u = [u_] + u_terms
+  r = [r_] + r_terms
+
+  logu_, logu_terms = _log_taylor((u_, ), (u_terms, ))
+
+  v_, v_terms = _bilinear_taylor_rule(lax.mul_p, (logu_, r_), (logu_terms, r_terms))
+
+  v_, v_terms = _exp_taylor((v_, ), (v_terms, ))
+
+  return v_, v_terms
```
If you're going to define things this way, I'd just call back into `jet`:
```suggestion
  return jet(lambda x, y: lax.exp(y * lax.log(x), primals_in, series_in)
```

duvenaud replied:
Haha, I feel silly now.  We started by trying to derive a giant monolithic formula for pow(x, y), but we already had all the ingredients!

jacobjinkelly replied:
@shoyer Thanks for pointing this out! I realized that I should change the `primal_out` propagated through the `exp` rule, so I incorporated your suggestion for propagation through `mul` and `log` in [77b7d9b](https://github.com/google/jax/pull/2652/commits/77b7d9be1924d4c73ab8884d9d9306245b6fbe21) (added you as a co-author)

jessebett replied:
@shoyer yeah! We're still unclear if there is a benefit from writing the expanded rules for primitives that can be composed of other primitives vs just calling jet on that composition like you have here.

shoyer commented:
I guess the more numerically stable but also more expensive alternate would be to recursively apply the JVP rule?

jacobjinkelly committed:
call jet and manually propagate through exp

Co-authored-by: Stephan Hoyer <shoyer@google.com>

jessebett commented:
@shoyer yeah you can see in the tests the function that looks like recursive calls to jvp:
https://github.com/google/jax/blob/c9c14c02d4c0e2ed6ebbc9bb4e6071a9d1c548fc/tests/jet_test.py#L33-L43

However this is significantly slower. Interesting to see if it's more numerically stable, though, because I'd expect the opposite. Not confident about intuition.

jacobjinkelly committed:
add finite test, add sep lims for binary_check

jacobjinkelly commented:
> I guess the more numerically stable but also more expensive alternate would be to recursively apply the JVP rule?

I think [77b7d9b](https://github.com/google/jax/pull/2652/commits/77b7d9be1924d4c73ab8884d9d9306245b6fbe21) fixes this. I changed the `primal_out` propagated through the `exp` rule to be `x ** y` (before it was `exp(y * log x)`.

I modified the tests in [c75855b](https://github.com/google/jax/pull/2652/commits/c75855b55c6b6df545ff5c82b66a2f219a7f90a1), to include the option to check jets on only finite values (by casting `np.inf`, `-np.inf`, and `np.nan` to `np.nan`). I also added separate lims (to keep the base `x` positive but allow the power `y` to be negative). With these modifications the tests for `jet(pow)` pass on a wide range of values.

shoyer commented:
Looks good to me!

@jacobjinkelly  That's kind of you to mark me as a co-author, but I think a one line review suggestion barely qualifies :)

@jessebett @duvenaud googlebot still needs your OK on this (`@googlebot I consent.`) before we can merge

duvenaud commented:
@googlebot I consent.

shoyer commented:
@jacobjinkelly could you kindly either remove my "co-author" status from the commit or use by google email instead (`shoyer@google.com`)?

jacobjinkelly commented:
@shoyer Ok rebased the relevant commit with your google email!

shoyer commented:
@jessebett this still needs your consent

jessebett commented:
@googlebot I consent.

shoyer commented:
Thanks everyone!

jacobjinkelly commented:
just to keep track this addresses #2431
